File: clojure/bob/bob.clj
(ns bob
  (:require [clojure.string :refer (blank? lower-case upper-case)]))

(defn- no-lower-case-letters? [text]
  (= (upper-case text) text))

(defn- contains-upper-case-letters? [text]
  (not= (lower-case text) text))

(def shouting? (every-pred no-lower-case-letters? contains-upper-case-letters?))
(def silence? blank?)

(defn- question? [text]
  (= \? (last text)))

(defn response-for [text]
  (cond
    (shouting? text) "Woah, chill out!"
    (silence? text) "Fine. Be that way!"
    (question? text) "Sure."
    :else "Whatever."))

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      silence?(input) ->
        "Fine. Be that way!"
      question?(input) ->
        "Sure."
      shouting?(input) ->
        "Whoa, chill out!"
      true ->
        "Whatever."
    end
  end

  defp silence?(input) do
    String.strip(input) == ""
  end

  defp shouting?(input) do
    String.upcase(input) == input && String.downcase(input) != input
  end

  defp question?(input) do
    String.ends_with?(input, "?")
  end
end
